common/deliver/deliver.go

```diff
@@ -260,6 +271,13 @@ func (h *Handler) deliverBlocks(ctx context.Context, srv *Server, envelope *cb.E
 
 	logger.Debugf("[channel: %s] Received seekInfo (%p) %v from %s", chdr.ChannelId, seekInfo, seekInfo, addr)
 
+	// Throttling limits the number of os threads created when getting blocks (file i/o's) from ledger.
+	if err := h.ThrottleSemaphore.Acquire(ctx); err != nil {
+		logger.Errorf("[channel: %s] Failed to acquire semaphore to get a block", chdr.ChannelId)
+		return cb.Status_INTERNAL_SERVER_ERROR, errors.WithMessage(err, "failed to acquire semaphore to get a block")
+	}
+	defer h.ThrottleSemaphore.Release()
```

Comment:
Again, this is for simplicity. There are multiple places calling ledger APIs. (1) chain.Reader().Iterator; (2) cursor.Next(); (3) srv.SendBlockResponse (only for deliver with pvtdata). Since we decide to throttle at the high level, I didn't break it into multiple locks.

Reply:
Yes, I agree to the trade-off and that's why its hard to make these choices without being backed by some performance measurement. To me, it would be like shooting in the dark.

Reply:
To solve a potential deadlock, I changed the code to call throttling multiple times, in each of the above 3 places calling ledger API. Also exported a WaitForNextBlock method from block iterator.